eZ/Publish/Core/QueryType/BuildIn/AbstractQueryType.php

```diff
@@ -0,0 +1,143 @@
+
+    /** @var \eZ\Publish\Core\MVC\ConfigResolverInterface */
+    protected $configResolver;
+
+    public function __construct(Repository $repository, ConfigResolverInterface $configResolver)
+    {
+        $this->repository = $repository;
+        $this->configResolver = $configResolver;
+    }
+
+    protected function configureOptions(OptionsResolver $resolver): void
+    {
+        $resolver->setDefaults([
+            'filter' => static function (OptionsResolver $resolver): void {
+                $resolver->setDefaults([
+                    'content_type' => [],
+                    'visible_only' => true,
+                ]);
+
+                $resolver->setAllowedTypes('content_type', 'array');
+                $resolver->setAllowedTypes('visible_only', 'bool');
+            },
+            'offset' => 0,
+            'limit' => self::DEFAULT_LIMIT,
+            'sort' => static function (OptionsResolver $resolver): void {
+                $resolver->setDefault('target', null);
+                $resolver->setAllowedTypes('target', ['null', 'string']);
+
+                $resolver->setDefault('direction', 'asc');
+                $resolver->setAllowedValues('direction', ['asc', 'desc']);
+                $resolver->setNormalizer('direction', function (Options $options, $direction) {
+                    if (strtolower($direction) === 'desc') {
+                        return Query::SORT_DESC;
+                    }
+
+                    return Query::SORT_ASC;
+                });
+
+                $resolver->setDefault('data', null);
+            },
+        ]);
+
+        $resolver->setAllowedTypes('offset', 'int');
+        $resolver->setAllowedTypes('limit', 'int');
+    }
+
+    abstract protected function getQueryFilter(array $parameters): Criterion;
+
+    protected function doGetQuery(array $parameters): Query
+    {
+        $query = new Query();
+        $query->filter = $this->buildFilters($parameters);
+
+        if ($parameters['sort']['target'] !== null) {
+            $query->sortClauses = $this->buildSortClauses(
+                $parameters['sort']['target'],
+                $parameters['sort']['direction'],
+                $parameters['sort']['data'] ?? []
+            );
+        }
+
+        $query->limit = $parameters['limit'];
+        $query->offset = $parameters['offset'];
+
+        return $query;
+    }
+
+    private function buildFilters(array $parameters): Criterion
+    {
+        $criteria = [
+            $this->getQueryFilter($parameters),
+        ];
+
+        if ($parameters['filter']['visible_only']) {
+            $criteria[] = new Visibility(Visibility::VISIBLE);
+        }
+
+        if (!empty($parameters['filter']['content_type'])) {
+            $criteria[] = new ContentTypeIdentifier($parameters['filter']['content_type']);
+        }
+
+        // Limit results to current SiteAccess tree root
+        $criteria[] = new Subtree($this->getRootLocationPathString());
```

Comment:
We must remember to document this. Also, do we need an option to disable that restriction ?

Reply:
> Also, do we need an option to disable that restriction ?

IMHO it's rare case, but I could imagine that someone would like to use build-in QT's in ezplatform-admin-ui extension. 

Added `siteaccess_aware` filter (enabled by default).